File: WebApi/godotrl/src/main/java/com/example/godotrl/cst/perception/CarDetection.java
package com.example.godotrl.cst.perception;

import br.unicamp.cst.core.entities.Codelet;
import br.unicamp.cst.core.entities.MemoryObject;
import com.example.godotrl.util.Updater;
import com.example.godotrl.util.Vector2;

import java.util.ArrayList;

public class CarDetection extends Codelet {
    MemoryObject visionMO = null;
    MemoryObject knownCarsMO = null;
    MemoryObject updateMO = null;

    @Override
    public void accessMemoryObjects() {
        visionMO = (MemoryObject) getInput("VISION");
        knownCarsMO = (MemoryObject) getOutput("KNOWN_CARS");
        updateMO = (MemoryObject) getOutput("UPDATE");
    }

    @Override
    public void calculateActivation() {

    }

    @Override
    public void proc() {
        if (((Updater) updateMO.getI()).canUpdateCarRead()) {
            ArrayList<Vector2> visionList = (ArrayList<Vector2>) visionMO.getI();
            ArrayList<Vector2> memoryList = (ArrayList<Vector2>) knownCarsMO.getI();

            for (Vector2 v : visionList) {
                if (!memoryList.contains(v)) {
                    memoryList.add(v);
                }
            }

            ((Updater) updateMO.getI()).updateCarRead();
        }
    }
}

File: GodotKotlin/MultiplatformFrogGame/bin/main/com/example/game/cst/perception/CarDetection.kt
package com.example.game.cst.perception

import br.unicamp.cst.core.entities.Codelet
import br.unicamp.cst.core.entities.MemoryObject
import com.example.game.godot.Car

/*
    Stores unknown cars in the visionMO in the long-term memory, defined by the knownCarsMO
*/

class CarDetection: Codelet() {
    var visionMO: MemoryObject? = null
    var knownCarsMO: MemoryObject? = null

    override fun accessMemoryObjects() {
        visionMO = getInput("VISION") as MemoryObject
        knownCarsMO = getOutput("KNOWN_CARS") as MemoryObject
    }

    override fun calculateActivation() {
        TODO("Not yet implemented")
    }

    override fun proc() {
        var visionList = visionMO?.i as ArrayList<Car>
        var memoryList = knownCarsMO?.i as ArrayList<Car>

        for (i in 0 until visionList.size) {
            if (visionList[i] !in memoryList) {
                memoryList.add(visionList[i])
            }
        }
    }
}